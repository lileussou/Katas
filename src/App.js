import { useCallback, useState, useEffect } from 'react';

import Rover from './objects/rover';
import Map from './objects/map';
import {changeRoverPosition} from './controller/roverMars';
import {Directions} from './models/directions';
import {Commands} from './models/commands';
import roverMarsImg from './image/roverMarsImg.png';
import mapMarsImg from './image/mapMarsImg.png';
import upArrow from './image/upArrow.png';
import downArrow from './image/downArrow.png';
import rightArrow from './image/rightArrow.png';
import leftArrow from './image/leftArrow.png';
import './App.css';

function App() {

  const mapMars = new Map(4,4);
  const [roverMars, setRoverMars] = useState(new Rover(1, 1, Directions.N));
  
  const moveRover = useCallback((command) => {
    const newRover = changeRoverPosition(roverMars, [command]);
    setRoverMars(new Rover(newRover.longitude, newRover.latitude, newRover.facing));
  }, [roverMars]);

  return (
    <>
      <div>
        <div className='parentDiv'>
          {mapMars.latitudeArray.map((latitudeElement, latitudeIndex) => (
            <div key = {latitudeIndex} className='latitude'> 
              {mapMars.longitudeArray.map((longitudeElement, longitudeIndex) => (
                <div key = {`${longitudeIndex}, ${latitudeIndex}`} className='longitude' id= {`${longitudeIndex}, ${latitudeIndex}`}>
                  {longitudeElement === roverMars.longitude
                    && latitudeElement === roverMars.latitude ?
                      <img alt='roverMarsImg' src={roverMarsImg} className='mapMarsImg'/>
                      :
                      <img alt='mapMarsImg' src={mapMarsImg} className='mapMarsImg'/>
                  } 
                </div>
              ))}
            </div>
          ))}
        </div>
        <div role='rover'>
          Position du Rover : {JSON.stringify(roverMars)}
        </div>
        <div role='marsMap'>
          Taille de la map : {JSON.stringify(mapMars.onlyCoordinates)}
        </div>
        <div className='arrowDiv'>
          <div>
            <img alt='upArrow' src={upArrow} className='arrows' onClick={() => moveRover(Commands.Forward)}/>
          </div>
          <div>
            <img alt='leftArrow' src={leftArrow} className='arrows' onClick={() => moveRover(Commands.Left)}/>
            <img alt='downArrow' src={downArrow} className='arrows' onClick={() => moveRover(Commands.BackWard)}/>
            <img alt='rightArrow' src={rightArrow} className='arrows' onClick={() => moveRover(Commands.Right)}/>
          </div>
        </div>
      </div>
    </>
  );
}

export default App;
