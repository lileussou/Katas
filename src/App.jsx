import { useCallback, useState } from 'react';

import Rover from './objects/rover';
import Map from './objects/map';
import {
  changeRoverPosition,
  angleBasedOnDirection,
} from './controller/roverMars';
import {
  createObstacle
} from './controller/mapMars';
import {Directions} from './models/directions';
import {Commands} from './models/commands';
import roverMarsImg from './image/roverMarsImg.png';
import mapMarsImg from './image/mapMarsImg.png';
import obstacleMarsImg from './image/obstacleMarsImg.png';
import upArrow from './image/upArrow.png';
import downArrow from './image/downArrow.png';
import rightArrow from './image/rightArrow.png';
import leftArrow from './image/leftArrow.png';
import './App.css';

function currentCase(mapMarsLongitude, mapMarsLatitude, roverMars, obstacleMars, angle) {
  if (mapMarsLongitude === roverMars.longitude
    && mapMarsLatitude === roverMars.latitude) {
      return  <img alt='roverMarsImg' src={roverMarsImg} className='mapMarsImg' style={{ transform: `rotate(${angle}deg)`}} />;
  } else if (obstacleMars.find((element) => element.longitude == mapMarsLongitude 
    && element.latitude == mapMarsLatitude)) {
      return <img alt='obstacleMarsImg' src={obstacleMarsImg} className='mapMarsImg'/>;
  } else {
    return <img alt='mapMarsImg' src={mapMarsImg} className='mapMarsImg'/>;
  }
}

function App() {

  const mapMars = new Map(4,4);
  const [roverMars, setRoverMars] = useState(new Rover(1, 1, Directions.N));
  const [angle, setAngle] = useState(0);
  const [currentObstacleLongitude, setCurrentObstacleLongitude] = useState('');
  const [currentObstacleLatitude, setCurrentObstacleLatitude] = useState('');
  const [obstacleList, setObstacleList] = useState([]);
  
  const moveRover = useCallback((command) => {
    const currentRoverMars = new Rover(roverMars.longitude, roverMars.latitude, roverMars.facing);
    const newRover = changeRoverPosition(roverMars, [command]);
    if (obstacleList.find((element) => element.longitude == newRover.longitude 
      && element.latitude == newRover.latitude)) {
      setRoverMars(currentRoverMars);
      window.alert("Le rover ne peux avancer en raison d'un obstacle");
    } else {
      setRoverMars(new Rover(newRover.longitude, newRover.latitude, newRover.facing));
      setAngle(angleBasedOnDirection(newRover.facing));
    }
  }, [roverMars, obstacleList]);

  const addObstacle = useCallback(() => {
    const newObstacle = createObstacle(currentObstacleLongitude, currentObstacleLatitude, mapMars, roverMars);
    setObstacleList(obstacleList => [...obstacleList, {longitude: newObstacle.longitude, latitude: newObstacle.latitude}]);
  }, [obstacleList, currentObstacleLongitude, currentObstacleLatitude, mapMars]);

  const deleteObstacle = useCallback((index) => {
    const newObstacleList = obstacleList.toSpliced(index, 1);
    console.log(newObstacleList);
    setObstacleList(newObstacleList);
  }, [obstacleList]);

  return (
    <>
      <div>
        <div>
          <div className='parentDiv'>
            {mapMars.latitudeArray.map((latitudeElement, latitudeIndex) => (
              <div key = {latitudeIndex} className='latitude'> 
                {mapMars.longitudeArray.map((longitudeElement, longitudeIndex) => (
                  <div key = {`${longitudeIndex}, ${latitudeIndex}`} className='longitude' id= {`${longitudeIndex}, ${latitudeIndex}`}>
                    {currentCase(longitudeElement, latitudeElement, roverMars, obstacleList, angle)} 
                  </div>
                ))}
              </div>
            ))}
          </div>
          <div role='rover'>
            Position du Rover : {JSON.stringify(roverMars)}
          </div>
          <div role='marsMap'>
            Taille de la map : {JSON.stringify(mapMars.onlyCoordinates)}
          </div>
        </div>
        <div>
          <label htmlFor="obstacleCoordinates">Longitude de l'obstacle:</label>
          <input type="number" id="obstacleLongitude" name="obstacleLongitude" min="1" max={`${mapMars.longitude}`} onChange={(e) => setCurrentObstacleLongitude(e.target.value)}/>

          <label htmlFor="obstacleCoordinates">Latitude de l'obstacle:</label>
          <input type="number" id="obstacleLatitude" name="obstacleLatitude" min="1" max={`${mapMars.latitude}`} onChange={(e) => setCurrentObstacleLatitude(e.target.value)}/>

          <input type="button" id="createObstacle" name="createObstacle" value="add" onClick={() => addObstacle()} />
        </div>
        <div>
          <ul>
            {obstacleList.map((obstacle, index) => (
              <li key = {`${obstacle} - ${index}`}>
                {JSON.stringify(obstacle)}
                <input type="button" id="deleteObstacle" name="deleteObstacle" value="x" onClick={() => deleteObstacle(index)} />
              </li>
            ))}
          </ul>
        </div>
        <div className='arrowDiv'>
          <div>
            <img alt='upArrow' src={upArrow} className='arrows' onClick={() => moveRover(Commands.Forward)}/>
          </div>
          <div>
            <img alt='leftArrow' src={leftArrow} className='arrows' onClick={() => moveRover(Commands.Left)}/>
            <img alt='downArrow' src={downArrow} className='arrows' onClick={() => moveRover(Commands.BackWard)}/>
            <img alt='rightArrow' src={rightArrow} className='arrows' onClick={() => moveRover(Commands.Right)}/>
          </div>
        </div>
      </div>
    </>
  );
}

export default App;
