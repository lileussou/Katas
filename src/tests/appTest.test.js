import {render, screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { act } from 'react-dom/test-utils';
import renderer from 'react-test-renderer';

import roverMarsImg from '../image/roverMarsImg.png';
import App from '../App';

describe('test roverMars comportment in DOM', () => {
    test('rover is at initial position', async () => {
        //ARRANGE
        render(<App />)
        //ACT
        await screen.findByRole('rover')
        //ASSERT
        expect(screen.getByRole('rover')).toHaveTextContent('{"longitude":1,"latitude":1,"facing":"N"}')
    })

    test('rover is at initial position with img', async () => {
        //ARRANGE
        render(<App />)
        //ACT
        const myRoverImg = screen.getByAltText('roverMarsImg')
        await screen.findByRole('rover')
        //ASSERT
        expect(screen.getByRole('rover')).toHaveTextContent('{"longitude":1,"latitude":1,"facing":"N"}')
        expect(myRoverImg.src).toContain(roverMarsImg)
    })

    test('rover is at initial position with img', async () => {
        //ARRANGE
        render(<App />)
        //ACT
        const myRoverImg = screen.getByAltText('roverMarsImg')
        await screen.findByRole('rover')
        //ASSERT
        expect(screen.getByRole('rover')).toHaveTextContent('{"longitude":1,"latitude":1,"facing":"N"}')
        expect(myRoverImg.src).toContain(roverMarsImg)
    })

    test('create a 4 x 4 map', async () => {
        //ARRANGE
        render(<App />)
        //ACT
        const myRoverImg = screen.getByAltText('roverMarsImg')
        await screen.findByRole('rover')
        await screen.findByRole('marsMap')
        //ASSERT
        expect(screen.getByRole('rover')).toHaveTextContent('{"longitude":1,"latitude":1,"facing":"N"}')
        expect(screen.getByRole('marsMap')).toHaveTextContent('{"longitude":4,"latitude":4}');
        expect(myRoverImg.src).toContain(roverMarsImg)
    })

    
    test('implement the 4 x 4 map and the rover', async () => {
        //ARRANGE
        render(<App />)
        //ACT
        const myMapImgs = screen.getAllByAltText('mapMarsImg');
        const myRoverImgsLength = screen.getAllByAltText('roverMarsImg').length;
        await screen.findByRole('rover')
        await screen.findByRole('marsMap')
        //ASSERT
        expect(screen.getByRole('rover')).toHaveTextContent('{"longitude":1,"latitude":1,"facing":"N"}')
        expect(screen.getByRole('marsMap')).toHaveTextContent('{"longitude":4,"latitude":4}');
        expect(myMapImgs).toHaveLength(16 - myRoverImgsLength);
    })

        
    test('rotate the rover to East', async () => {
        //ARRANGE
        act(() => {
            render(<App />)
        })
        //ACT
        await screen.findByRole('rover')
        act(() => {
            userEvent.click(screen.getByAltText('rightArrow'))
        })
        //ASSERT
        expect(screen.getByRole('rover')).toHaveTextContent('{"longitude":1,"latitude":1,"facing":"E"}')
    })

    test('move the rover to move forward', async () => {
        //ARRANGE
        act(() => {
            render(<App />)
        })
        //ACT
        await screen.findByRole('rover')
        act(() => {
            userEvent.click(screen.getByAltText('upArrow'))
        })
        //ASSERT
        expect(screen.getByRole('rover')).toHaveTextContent('{"longitude":1,"latitude":2,"facing":"N"}')
    })

    test('First snapshot Test', () => {
        const component = renderer.create(
            <App />
        );

        let tree = component.toJSON();

        expect(tree).toMatchSnapshot(); 
    })
})