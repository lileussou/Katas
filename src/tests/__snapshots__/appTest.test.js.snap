// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test roverMars comportment in DOM First snapshot Test 1`] = `
<div>
  <div>
    <div
      className="parentDiv"
    >
      <div
        className="latitude"
      >
        <div
          className="longitude"
          id="0, 0"
        >
          <img
            alt="roverMarsImg"
            className="mapMarsImg"
            src="roverMarsImg.png"
            style={
              Object {
                "transform": "rotate(0deg)",
              }
            }
          />
        </div>
        <div
          className="longitude"
          id="1, 0"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
        <div
          className="longitude"
          id="2, 0"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
        <div
          className="longitude"
          id="3, 0"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
      </div>
      <div
        className="latitude"
      >
        <div
          className="longitude"
          id="0, 1"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
        <div
          className="longitude"
          id="1, 1"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
        <div
          className="longitude"
          id="2, 1"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
        <div
          className="longitude"
          id="3, 1"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
      </div>
      <div
        className="latitude"
      >
        <div
          className="longitude"
          id="0, 2"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
        <div
          className="longitude"
          id="1, 2"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
        <div
          className="longitude"
          id="2, 2"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
        <div
          className="longitude"
          id="3, 2"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
      </div>
      <div
        className="latitude"
      >
        <div
          className="longitude"
          id="0, 3"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
        <div
          className="longitude"
          id="1, 3"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
        <div
          className="longitude"
          id="2, 3"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
        <div
          className="longitude"
          id="3, 3"
        >
          <img
            alt="mapMarsImg"
            className="mapMarsImg"
            src="mapMarsImg.png"
          />
        </div>
      </div>
    </div>
    <div
      role="rover"
    >
      Position du Rover : 
      {"longitude":1,"latitude":1,"facing":"N"}
    </div>
    <div
      role="marsMap"
    >
      Taille de la map : 
      {"longitude":4,"latitude":4}
    </div>
  </div>
  <div>
    <label
      htmlFor="obstacleCoordinates"
    >
      Longitude de l'obstacle:
    </label>
    <input
      id="obstacleLongitude"
      max="4"
      min="1"
      name="obstacleLongitude"
      onChange={[Function]}
      type="number"
    />
    <label
      htmlFor="obstacleCoordinates"
    >
      Latitude de l'obstacle:
    </label>
    <input
      id="obstacleLatitude"
      max="4"
      min="1"
      name="obstacleLatitude"
      onChange={[Function]}
      type="number"
    />
    <input
      id="createObstacle"
      name="createObstacle"
      onClick={[Function]}
      type="button"
      value="add"
    />
  </div>
  <div>
    <ul />
  </div>
  <div
    className="arrowDiv"
  >
    <div>
      <img
        alt="upArrow"
        className="arrows"
        onClick={[Function]}
        src="upArrow.png"
      />
    </div>
    <div>
      <img
        alt="leftArrow"
        className="arrows"
        onClick={[Function]}
        src="leftArrow.png"
      />
      <img
        alt="downArrow"
        className="arrows"
        onClick={[Function]}
        src="downArrow.png"
      />
      <img
        alt="rightArrow"
        className="arrows"
        onClick={[Function]}
        src="rightArrow.png"
      />
    </div>
  </div>
</div>
`;
